### PR TITLE
unify joining and shard group link code

### PR DESCRIPTION
**Pull request by sjpotter**
we know have a common set of connection handling code

+ python test string formating fix

### CODE REVIEW

File: redisraft.h [hunk 1]
```diff
@@ -553,6 +553,9 @@ typedef struct SnapshotResult {
     char err[256];
 } SnapshotResult;
 
+extern char * join_type;
+extern char * link_type;
+
```
Comment (yossigo) [inline redisraft.h:558]: I think there should be a better way to handle this than introduce these globals. If we're trying to keep this reusable and generic, maybe we should just have a callback function for unrecovarable error handling - which will be responsible for producing the apropriate error reply as well perform any other specific cleanup tasks we may need in the future.

File: common.c [hunk 2]
```diff
@@ -177,4 +177,65 @@ bool parseMovedReply(const char *str, NodeAddr *addr)
...
+/* future types should remain 4 characters or stack array in common.c/joinLinkIdleCallback() has to expand */
+char * join_type = "join";
+char * link_type = "link";
+
+/* Invoked when the connection is not connected or actively attempting
+ * a connection.
+ */
+void joinLinkIdleCallback(Connection *conn)
+{
+    char err_msg[42];
```
Comment (yossigo) [inline common.c:191]: I'd make it just a bit bigger to accomodate for future error message changes, etc.

File: common.c [hunk 3]
```diff
@@ -177,4 +177,65 @@ bool parseMovedReply(const char *str, NodeAddr *addr)
...
+
+    time_t now;
+    time(&now);
+
+    if (state->failed) {
+        LOG_ERROR("Cluster %s: unrecoverable error, check logs", state->type);
+        goto exit_fail;
+    }
+
+    if (difftime(now, state->start) > rr->config->join_timeout) {
```
Comment (yossigo) [inline common.c:204]: We need to either have a different timeout setting or document that the join applies to shardgroup linking as well.

File: common.c [hunk 4]
```diff
@@ -177,4 +177,65 @@ bool parseMovedReply(const char *str, NodeAddr *addr)
...
+    /* Establish connection. We silently ignore errors here as we'll
+     * just get iterated again in the future.
+     */
+    ConnConnect(state->conn, &state->addr_iter->addr, state->connect_callback);
+    return;
+
+exit_fail:
+    ConnAsyncTerminate(conn);
+
+    snprintf(err_msg, 42, "ERR failed to %s cluster, please check logs", state->type);
```
Comment (yossigo) [inline common.c:228]:
```suggestion
    snprintf(err_msg, sizeof(err_msg), "ERR failed to %s cluster, please check logs", state->type);
```

Safer this way.

File: redisraft.h [hunk 5]
```diff
@@ -580,7 +583,20 @@ static int inline checkInRedisModuleCall(void) {
     return redis_raft_in_rm_call;
 }
 
+typedef struct JoinLinkState {
```
Comment (yossigo) [inline redisraft.h:586]: Now that's it's totally generic, maybe this should be named something like `NodeExecuteCommandState` or something similar?

File: common.c (hunk 3 above)
Comment (sjpotter) [reply common.c:204]: I think documenting is fine, they are both "join" type operations

File: common.c (hunk 2 above)
Comment (sjpotter) [reply common.c:191]: I guess its not the answer as is :)

File: redisraft.h (hunk 5 above)
Comment (sjpotter) [reply redisraft.h:586]: that's less descriptive to me.  i.e. this is state that relates to joining/linking redis raft nodes/clusters to each other.    I'm unsure what nodeExecuteCommand actually means

File: redisraft.h (hunk 1 above)
Comment (sjpotter) [reply redisraft.h:558]: ok, can get rid of the globals, I'd rather not add another callback yet if not understood what it really adds over the existign code.  I'll just add allocated data to state object and free in its normal callback

**Review by yossigo — APPROVED** [review]

**Pull request merged**